### PR TITLE
feat(realtime): add "enabled" toggle for audio streams

### PR DESCRIPTION
**Pull request by alexjurkiewicz**
## Summary

- Add an `enabled` flag to realtime stream config so a configured stream can stay in settings while being removed from the active runtime set
- Disabled streams are filtered out of the desired source set during stream reconfigure, so toggling `enabled` off removes the running source and toggling it back on re-adds it
- Legacy configs predate the field, so missing `enabled` is normalized to `true` before config unmarshal and persisted back to disk
- Quiet-hours and other active-stream loops now ignore disabled streams so they cannot be restarted by scheduler logic or treated as active capture sources
- Settings UI exposes an `Enabled` checkbox per stream and renders disabled streams as a distinct state

## Summary by CodeRabbit

* **New Features**
  * Per-stream enable/disable toggles in view and edit modes; disabled streams show distinct styling and status pill and have a read-only checkbox when controls are disabled.
* **Bug Fixes**
  * Disabled streams are excluded from audio processing, monitoring, scheduling, level reports, and are removed on reconfiguration; toggling now triggers proper reconfiguration and cleanup.
* **Tests**
  * Added/updated tests covering enable/disable behavior and settings coercion.
* **Chores**
  * Settings coercion and migration ensure legacy streams default to enabled for compatibility.

### CODE REVIEW

**Comment by alexjurkiewicz** [comment]
Hi, this is my first contribution to the repo. It's mostly coded with Claude.

Use-case: I was debugging my RTSP source (and Atom Echo) and found myself having to repeatedly add/remove it from Birdnet-Go as I was testing things locally and solving some crash bugs. Needing to re-add the stream each time was very painful, and I wished for this feature. So, here it is!

File: frontend/src/lib/desktop/components/forms/StreamCard.svelte [hunk 1]
```diff
@@ -218,6 +229,8 @@
         return t('settings.audio.streams.status.connecting');
       case 'error':
         return t('settings.audio.streams.status.error');
+      case 'disabled':
+        return 'Disabled';
```
[the remark this answers is not in the document]
Comment (alexjurkiewicz) [reply frontend/src/lib/desktop/components/forms/StreamCard.svelte]: Fixed — rather than adding a new key, we reuse the existing `t('settings.alerts.disabled')` which is already in all locale files. This avoids duplicating identical strings.

File: frontend/src/lib/desktop/components/forms/StreamCard.svelte [hunk 2]
```diff
@@ -409,6 +430,14 @@
           {/if}
         </div>
 
+        <Checkbox
+          checked={editEnabled}
+          onchange={checked => (editEnabled = checked)}
+          label="Enabled"
```
[the remark this answers is not in the document]
Comment (alexjurkiewicz) [reply frontend/src/lib/desktop/components/forms/StreamCard.svelte]: Fixed — reusing the existing `t('settings.alerts.enabled')` key rather than a new stream-specific one, since the text is identical and the key already exists across all locales.

File: frontend/src/lib/desktop/components/forms/StreamCard.svelte [hunk 3]
```diff
@@ -491,6 +520,16 @@
           <p class="text-xs text-[var(--color-base-content)]/70 mt-1.5 mb-1">
             Total: {totalData} • Rate: {currentRate} • Last: {lastDataAgo}
           </p>
+
+          <div class="mt-2">
+            <Checkbox
+              checked={stream.enabled}
+              onchange={updateEnabled}
+              label="Enabled"
```
[the remark this answers is not in the document]
Comment (alexjurkiewicz) [reply frontend/src/lib/desktop/components/forms/StreamCard.svelte]: Fixed — same as above, using `t('settings.alerts.enabled')` to avoid duplicating an already-translated string.

File: frontend/src/lib/desktop/components/forms/StreamManager.svelte [hunk 4]
```diff
@@ -515,6 +526,13 @@
                 size="sm"
               />
             {/if}
+            {#if healthySummary.disabled > 0}
+              <StatusPill
+                variant="neutral"
+                label="{healthySummary.disabled} Disabled"
```
[the remark this answers is not in the document]
Comment (alexjurkiewicz) [reply frontend/src/lib/desktop/components/forms/StreamManager.svelte]: Fixed — using `t('settings.alerts.disabled')` (already present in all locale files) instead of adding a new key.

File: internal/analysis/audio_pipeline_service.go [hunk 5]
```diff
@@ -904,7 +904,7 @@ func (p *AudioPipelineService) buildSourceConfigsWithModels() []sourceConfigWith
 	// RTSP streams.
 	for i := range settings.Realtime.RTSP.Streams {
 		stream := &settings.Realtime.RTSP.Streams[i]
-		if stream.URL == "" {
+		if stream.URL == "" || !stream.IsEnabled() {
 			continue
 		}
```
[the remark this answers is not in the document]
Comment (alexjurkiewicz) [reply internal/analysis/audio_pipeline_service.go]: Fixed — removed the `if len(allActiveIDs) > 0` guard so `UpdateMonitors` is always called with the current desired map (including an empty one), ensuring stale monitors are torn down when the last stream is disabled.

File: internal/analysis/audio_pipeline_stream_enabled_test.go [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+			Models:  []string{"birdnet"},
+		},
+	}
+	conf.SetTestSettings(settings)
+
+	p := &AudioPipelineService{}
+	configs := p.buildSourceConfigsWithModels()
+
+	assert.Len(t, configs, 1, "disabled streams should be excluded from active source configs")
+	assert.Equal(t, "rtsp://cam1", configs[0].config.ConnectionString)
```
[the remark this answers is not in the document]
Comment (alexjurkiewicz) [reply internal/analysis/audio_pipeline_stream_enabled_test.go]: Fixed — changed to `require.Len` so the test aborts before reaching the `configs[0]` index if the length assertion fails.

File: internal/conf/stream_persistence_test.go [hunk 7]
```diff
@@ -268,6 +285,113 @@ transport: "udp"
...
+        transport: "tcp"
+`
+	err := os.WriteFile(configPath, []byte(configYAML), 0o600)
+	require.NoError(t, err)
+
+	oldPath := ConfigPath
+	t.Cleanup(func() {
+		ConfigPath = oldPath
+		viper.Reset()
+	})
```
[the remark this answers is not in the document]
Comment (alexjurkiewicz) [reply internal/conf/stream_persistence_test.go:399]: Fixed — snapshot the global settings before `Load()` and restore it in `t.Cleanup` alongside the existing `ConfigPath`/`viper.Reset()` so later tests don't inherit the temporary config.

File: internal/analysis/audio_pipeline_stream_enabled_test.go [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+	})
+	require.NoError(t, err)
+
+	_, exists := engine.Registry().GetByConnection(connection)
+	require.True(t, exists, "test setup should start with the stream registered")
+
+	// bufferMgr is intentionally nil: UpdateMonitors is only called when
+	// len(allActiveIDs) > 0, and this scenario ends with zero active streams.
+	p := &AudioPipelineService{engine: engine}
+	p.reconfigureChangedSources(make(chan audiocore.AudioLevelData))
```
[the remark this answers is not in the document]
Comment (alexjurkiewicz) [reply internal/analysis/audio_pipeline_stream_enabled_test.go]: Valid bug — fixed by replacing the unconditional call with a `if p.bufferMgr != nil` guard (consistent with the existing nil check at line 304 of the same file). The "always reconcile" intent from the CodeRabbit fix is preserved for production code paths where `bufferMgr` is always initialised, while test cases with a nil manager are no longer affected. Updated the test comment to reflect the new guard.

File: internal/api/v2/audio_level.go [hunk 9]
```diff
@@ -546,6 +552,26 @@ func (c *Controller) isSourceInactive(
...
+		if _, ok := registry.Get(sourceID); !ok {
+			delete(levels, sourceID)
+			delete(lastUpdateTime, sourceID)
+			delete(lastNonZeroTime, sourceID)
+			pruned = true
+		}
+	}
+	return pruned
+}
+
```
Comment (alexjurkiewicz) [inline internal/api/v2/audio_level.go:570]: Worth noting as a follow-up: `pruneRemovedSources`, `checkSourceActivity`, and `updateAudioLevel` all have to touch the same three parallel maps (`levels`, `lastUpdateTime`, `lastNonZeroTime`) in lockstep. Consolidating those into a single `map[string]sourceState` struct would make all three functions simpler and eliminate the risk of them getting out of sync in future. Not a blocker for this PR though.

File: internal/conf/config.go [hunk 10]
```diff
@@ -547,12 +548,21 @@ const DefaultTransport = "tcp"
...
 	Models     []string         `yaml:"models,omitempty" json:"models,omitempty" mapstructure:"models"` // Model IDs for this stream (e.g., ["birdnet", "perch_v2"])
 }
 
+// IsEnabled returns the effective enabled state for a stream.
+// A nil receiver returns true so callers can safely use range-index pointers
+// without a separate nil guard; a missing stream is treated as enabled rather
+// than silently dropped.
+func (s *StreamConfig) IsEnabled() bool {
+	return s == nil || s.Enabled
+}
```
[the remark this answers is not in the document]
Comment (alexjurkiewicz) [reply internal/conf/config.go:561]: False positive. `validateStreamsSection` only decodes into a temporary struct for validation — it is never written back to settings. The actual write path goes through `mergeJSONIntoStruct` → `deepMergeMaps`, which starts from the current settings state (serialised to a map via `json.Marshal`) and only overwrites fields that are explicitly present in the incoming JSON payload. If `enabled` is absent, `deepMergeMaps` preserves the existing value. The only edge case is a brand-new stream being added, but the frontend always sends `enabled: true` explicitly when creating a stream.

File: internal/api/v2/audio_level.go [hunk 11]
```diff
@@ -412,8 +415,7 @@ func (c *Controller) getAudioCardSources(registry *audiocore.SourceRegistry) []*
 
 // addStreamSourcesToLevels adds all configured stream sources to the levels map.
 func (c *Controller) addStreamSourcesToLevels(registry *audiocore.SourceRegistry, levels map[string]audiocore.AudioLevelData, isAuthenticated bool) {
-	for i := range c.Settings.Realtime.RTSP.Streams {
-		stream := &c.Settings.Realtime.RTSP.Streams[i]
+	for i, stream := range c.Settings.Realtime.RTSP.EnabledStreams() {
 		source, ok := registry.GetByConnection(stream.URL)
 		if !ok {
 			continue
```
[the remark this answers is not in the document]
Comment (alexjurkiewicz) [reply internal/api/v2/audio_level.go:421]: The names used by cameras are stable opaque identifiers. It's OK if numbers are missing.

A sequential counter has a worse problem: if the user re-enables the previously-disabled stream, names shuffle — what was `camera-1` (the second stream) becomes `camera-2`. The current index-based approach at least gives *stable* names: `camera-2` always refers to stream[1] regardless of what else is enabled or disabled. The right long-term fix is a stable hash derived from the stream URL, but that's out of scope here. Leaving as-is.

**Comment by alexjurkiewicz** [comment]
**Re: CodeRabbit review 4145071865 — outside-diff comments**

**`audio_level.go:418-426` — camera naming gaps**

Same issue as https://github.com/tphakala/birdnet-go/pull/2836#discussion_r3115154394

A sequential counter trades gaps for instability: if the user re-enables a previously-disabled stream, all the names shift (what was `camera-1` becomes `camera-2`, etc.). The current index-based approach gives *stable* names — `camera-2` always means stream[1] regardless of enable/disable state. The proper fix is a stable hash derived from the stream URL, which is out of scope for this PR. Leaving as-is.

**`audio_level.go:323-345` — stale updates re-entering `levels` after prune**

Not a real race. The event loop is a single goroutine — `subscriberChan` and `activityCheck.C` are handled in the same `select`, so `updateAudioLevel` and `pruneRemovedSources` can never run concurrently. The only scenario is a stale packet arriving in `subscriberChan` after a prune in the same `select` iteration, but the audio pipeline tears the source down before the prune runs, so no further packets can arrive for it. If one did slip through, `checkSourceActivity` would zero it out on the next 1s tick. Adding a per-packet registry lookup is not worth the overhead for this theoretical edge case.

**Comment by tphakala** [comment]
Thanks for this contribution, nice feature to have. The overall design is solid, especially the `AllStreams()`/`EnabledStreams()` iterator approach and the backward-compatible migration for legacy configs.

A few things need attention before this can be merged though.

CI is currently failing on three checks. `StreamCard.svelte` needs a Prettier pass. In the test file `audio_pipeline_stream_enabled_test.go` at line 62, `context.Background()` should be `t.Context()` since we require Go 1.24+ patterns (gocritic catches this). And the `Load()` function in `config.go` hit cyclomatic complexity 32, which is over our threshold of 30. The stream-enabled migration block pushed it over the edge, so it probably needs to be extracted into a small helper function.

One design question: `IsEnabled()` returns `true` for a nil receiver. The comment explains the rationale, but in practice no call site actually needs this since `EnabledStreams()` already provides non-nil pointers. A nil `StreamConfig` silently acting as "enabled" could mask bugs rather than surfacing them. I'd prefer either dropping the nil guard entirely or flipping it so nil means disabled.

For the translation keys, reusing `settings.alerts.enabled`/`settings.alerts.disabled` works for now, but these are semantically about alert toggles. Some languages might translate "enabled" differently depending on context (adjective vs state label). Not a blocker, just something to be aware of.

Everything else looks good to me. The quiet hours interaction is handled correctly, the monitor teardown fix for the last-stream-disabled case is a nice catch, and the test coverage is thorough. Looking forward to seeing the CI fixes.

**Comment by alexjurkiewicz** [comment]
Thank you for the feedback! I'll be honest with you, I wasn't really thinking about the multi-stream cases while writing this PR (I have only one microphone), so I'm glad the agents figured out the monitor teardown and api sections specifically!!

I am hoping to follow this up with a few other improvements:
* The UI on the stream config page for failed connections showing a timeline of connection/disconnection/backoff doesn't seem to render properly for me, it keeps appearing and disappearing. I will investigate this
* Add persistent logging for streams (you can sort-of reconstruct what happened by reading notifications, but it's painful) (might be obsoleted when the visual timeline works)
* More detailed stream latency / jitter monitoring

**Comment by alexjurkiewicz** [comment]
> One design question: IsEnabled() returns true for a nil receiver. The comment explains the rationale, but in practice no call site actually needs this since EnabledStreams() already provides non-nil pointers. A nil StreamConfig silently acting as "enabled" could mask bugs rather than surfacing them. I'd prefer either dropping the nil guard entirely or flipping it so nil means disabled.

I'm happy to follow your preference.

> For the translation keys, reusing settings.alerts.enabled/settings.alerts.disabled works for now, but these are semantically about alert toggles. Some languages might translate "enabled" differently depending on context (adjective vs state label). Not a blocker, just something to be aware of.

Wow, great catch. Codex suggests that Spanish and Polish will have this issue:

```
- settings.alerts.enabled = Activado
- settings.alerts.disabled = Desactivado
- settings.audio.streams.streamLabel = Transmisión

Transmisión is feminine, so if that alert key is reused for a stream, the
natural form would be Activada / Desactivada, not Activado / Desactivado.
```

I'll add a new key with machine translated values for non-English languages (Codex thinks the translation differs only for Spanish and Polish). Would you prefer the non-English translation strings to be blank or omitted? I'm not that familiar with i18n...

File: internal/conf/config.go [hunk 12]
```diff
@@ -2150,6 +2226,45 @@ func (s *Settings) MigrateSourceModels() bool {
...
+	for i, rawStream := range streams {
+		streamMap, ok := rawStream.(map[string]any)
+		if !ok {
+			normalized[i] = rawStream
+			continue
+		}
+		if _, exists := streamMap["enabled"]; exists {
+			normalized[i] = rawStream
+			continue
+		}
```
[the remark this answers is not in the document]
Comment (tphakala) [reply internal/conf/config.go]: Fixed in b0ae133d - also added a test for the nil case in TestNormalizeRTSPStreamEnabledDefaults_NilValue.

File: internal/conf/validate.go [hunk 13]
```diff
@@ -314,8 +313,7 @@ func (r *RTSPSettings) ValidateStreams() error {
 	names := make(map[string]bool)
 	urls := make(map[string]bool)
 
-	for i := range r.Streams {
-		stream := &r.Streams[i]
+	for i, stream := range r.AllStreams() {
 		// Validate individual stream
 		if err := stream.Validate(); err != nil {
 			return fmt.Errorf("stream %d: %w", i+1, err)
```
[the remark this answers is not in the document]
Comment (tphakala) [reply internal/conf/validate.go:319]: Intentional design: ValidateStreams() uses AllStreams() to catch config errors early, even in disabled streams. Skipping validation for disabled streams would let broken configs hide and fail at runtime when re-enabled. ApplyStreamDefaults() follows the same pattern - all config-time operations validate everything, only runtime paths filter to enabled.

File: internal/api/v2/audio_level.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (tphakala) [reply internal/api/v2/audio_level.go:421]: Agreed with the PR author's analysis in the earlier thread: index-based naming keeps camera names stable across enable/disable toggles. A sequential counter would shuffle names when a previously disabled stream is re-enabled.

File: internal/conf/config.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (alexjurkiewicz) [reply internal/conf/config.go]: Why would someone set `enabled: null` ..?

**Comment by tphakala** [comment]
Hi @alexjurkiewicz, a licensing note about your contribution here.

I'm planning to relicense BirdNET-Go from CC BY-NC-SA 4.0 to AGPL-3.0, an OSI-approved open source license. It would keep the copyleft, share-alike spirit but remove the non-commercial restriction and make the project properly open source. There are no plans to commercialize the project; this is to align better with the open source community and make packaging easier. Full context is in #4243, and consent is tracked in #4244.

Since your work is part of the codebase, I'd like your agreement before relicensing. Could you reply confirming:

> I agree that my contributions to birdnet-go, past and future, may be licensed under AGPL-3.0.

Thanks a lot for this contribution.

**Comment by alexjurkiewicz** [comment]
Good luck ! I have done relicensing before and it is a pain!!!

I agree that my contributions to birdnet-go, past and future, may be
licensed under AGPL-3.0.
